a11y: remove SemanticsSortOrder; sort locally only; semanticsOwner post-test check

yjbanov opened the pull request:
This PR prepares us for the upcoming a11y traversal changes. Namely:

- Remove `SemanticsSortOrder` - we now rely purely on `SemanticsSortKey` for sorting.
- Sorting is done only locally; we do not sort across branches.
- Add a post-test check that all `SemanticsHandle`s are disposed of. This prevents cross-test contamination. Ever wondered why in some tests the `id` starts with 1 and in others it starts with 2? Now, if you use `debugResetSemanticsIdCounter` before your test, all `id`s are _guaranteed_ to start with 1 (the root node `id` remains to be hard-coded to 0).

Breaking change announcement: https://groups.google.com/forum/#!topic/flutter-dev/iCoLnW31heE

yjbanov committed:
a11y: remove SemanticsSortOrder; sort locally only; semanticsOwner post-test check

goderbauer commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -3276,17 +3276,17 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
+  /// Sets the [SemanticsNode.sortKey] to the given value.
   ///
-  /// This defines how this node will be sorted with the other semantics nodes
+  /// This defines how this node are sorted with the other semantics nodes
```
mention that this is only a local sort amongst siblings?

yjbanov replied:
Done.

goderbauer commented on packages/flutter/test/widgets/custom_painter_test.dart:
```diff
@@ -357,13 +357,13 @@ void _defineTests() {
           id: 1,
           previousNodeId: -1,
-          nextNodeId: expectedId,
+          nextNodeId: -1,
```
This looks odd now. Does -1 mean "not specified"? Can we make this the default so you don't have to write it out in your tests?

yjbanov replied:
Done (see https://github.com/flutter/flutter/pull/15537/commits/b2e2aa0c6b2eb4bfe51ba1096e861ed919b39a34)

yjbanov committed:
update accessibility test framework

- default nextNodeId/previousNodeId to -1
- stop treating null as opt-out from value testing
- add `id`, `TestSemantics.root`, and `tags` to the suggested code in the TestSemantics failure message
- fix a small bug with raw string escaping
- update all tests accordingly

yjbanov committed:
fix sortKey doc

yjbanov commented:
Made the requested test framework updates. PTAL.

goderbauer reviewed (approved):
LGTM

@gspencergoog can you take a look as well since you know this code the best?

gspencergoog commented on packages/flutter/test/widgets/semantics_tester.dart:
```diff
@@ -350,6 +354,14 @@ class SemanticsTester {
+    // performed irrespective of whether the owner was created via
+    // SemanticsTester or directly. When the test succeeds, this tear-down
+    // becomes a no-op.
+    addTearDown(dispose);
```
Good find!  I have been meaning to check on this because it causes extra failures when a test fails, making it harder to see the real failure sometimes.

gspencergoog commented:
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/34173247-ae854d5c-e4a9-11e7-9f19-73b35006f6f8.png)

This looks really good, Yegor.  Thanks for all of your work here.

yjbanov committed:
prefer const over final
